pkg/runtime/handler.go

```diff
@@ -0,0 +1,167 @@
+	return &handlerFactory{
+		tmplRepo:          tmplRepo,
+		attrDescFinder:    df,
+		typeChecker:       expr,
+		infrdTypsCache:    make(map[string]proto.Message),
+		builderInfoFinder: builderInfoFinder,
+	}
+}
+
+// Build instantiates a Handler object using the passed in handler and instances configuration.
+func (h *handlerFactory) Build(handler *pb.Handler, instances []*pb.Instance, env adapter.Env) (adapter.Handler, error) {
+	infrdTypsByTmpl, err := h.inferTypesGrpdByTmpl(instances)
+	if err != nil {
+		return nil, err
+	}
+
+	// HandlerBuilder should always be present for a valid configuration (reference integrity should already be checked).
+	hndlrBldrInfo, _ := h.builderInfoFinder(handler.Adapter)
+
+	hndlrBldr := hndlrBldrInfo.CreateHandlerBuilder()
+	if hndlrBldr == nil {
+		msg := fmt.Sprintf("nil HandlerBuilder instantiated for adapter '%s' in handler config '%s'", handler.Adapter, handler.Name)
+		glog.Warning(msg)
+		return nil, fmt.Errorf(msg)
+	}
+
+	var hndlr adapter.Handler
+	hndlr, err = h.build(hndlrBldr, infrdTypsByTmpl, handler.Params, env)
+	if err != nil {
+		msg := fmt.Sprintf("cannot configure adapter '%s' in handler config '%s': %v", handler.Adapter, handler.Name, err)
+		glog.Warning(msg)
+		return nil, fmt.Errorf(msg)
+	}
+
+	return hndlr, err
+}
+
+func (h *handlerFactory) build(hndlrBldr adapter.HandlerBuilder, infrdTypesByTmpl map[string]types,
+	adapterCnfg interface{}, env adapter.Env) (handler adapter.Handler, err error) {
+	// calls into handler can panic. If that happens, we will log and return error with nil handler
+	defer func() {
+		if r := recover(); r != nil {
+			msg := fmt.Sprintf("handler panicked with '%v' when trying to configure the associated adapter."+
+				" Please remove the handler or fix the configuration.", r)
+			glog.Warningf(msg)
+			handler = nil
+			err = fmt.Errorf(msg)
+			return
+		}
+	}()
+
+	for tmplName, typs := range infrdTypesByTmpl {
+		// ti should be there for a valid configuration.
+		ti, _ := h.tmplRepo.GetTemplateInfo(tmplName)
+		if err := ti.ConfigureType(typs, &hndlrBldr); err != nil {
+			return nil, fmt.Errorf("cannot configure handler types %v for mesh function name '%s': %v", typs, tmplName, err)
+		}
+	}
+
+	return hndlrBldr.Build(adapterCnfg.(proto.Message), env)
+}
+
+func (h *handlerFactory) inferTypesGrpdByTmpl(instances []*pb.Instance) (map[string]types, error) {
+	infrdTypesByTmpl := make(map[string]types)
+	for _, instance := range instances {
+		infrdType, err := h.inferType(instance)
+		if err != nil {
+			return infrdTypesByTmpl, err
+		}
+
+		if _, exists := infrdTypesByTmpl[instance.GetTemplate()]; !exists {
+			infrdTypesByTmpl[instance.GetTemplate()] = make(types)
+		}
+
+		infrdTypesByTmpl[instance.GetTemplate()][instance.Name] = infrdType
+	}
+	return infrdTypesByTmpl, nil
+}
+
+func (h *handlerFactory) inferType(instance *pb.Instance) (proto.Message, error) {
+
+	var infrdType proto.Message
+	var err error
+	var found bool
+
+	h.lock.RLock()
```

Comment:
Please refer to comment above.. I would imagine this code to be totally usable across multiple go routines.. any reason to not do that ?